### PR TITLE
Set Up Simulator Datatypes

### PR DESCRIPTION
**Pull request by intermezzio**
- Add Simulator class to exports.
- Add scheduled strategy class
- Rename strategy class
- Fix imports
- Add Money and refactor Portfolio class to support free cash in all assets.
- Add tests for the Money class

# Description

Refactor the backtesting classes to support custom types in preparation for implementing
the simulation.

Fixes # (issue)

## Type of change
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

Unit tests

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [  ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/coin_test/util/money.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
+"""Initializes the Money class."""
+
+from .ticker import Ticker
+
+
+class Money:
+    """Amount of money in a given currency."""
```
Comment (teadetime) [inline src/coin_test/util/money.py:7]: need a different word than Money to explain this

File: src/coin_test/util/money.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        self.currency = currency
+        self.qty = qty
+
+    def _can_compare(self, other: "Money") -> None:
+        """Check if Money objects can be compared."""
+        if not isinstance(other, Money):
+            raise NotImplementedError
+
+        if self.currency != other.currency:
+            raise ValueError("Can't compare equality across different tickers")
```
Comment (teadetime) [inline src/coin_test/util/money.py:25]: I think I threw a different error in Ticker, we should standardize

File: src/coin_test/backtest/portfolio.py [hunk 3]
```diff
@@ -1,21 +1,46 @@
...
+            base_currency: The base currency of the portfolio
+            assets: A dictionary of all holdings
+
+        Raises:
+            ValueError: If the parameters are not inputted properly
         """
-        self.cash = cash
+        self.base_currency = base_currency
         self.assets = assets if assets is not None else {}
+        self.reserved = {a: Money(a, 0) for a in assets}
```
Comment (teadetime) [inline src/coin_test/backtest/portfolio.py:21]: Docs on attributes would be helpful

File: tests/backtest/test_portfolio.py [hunk 4]
```diff
@@ -1,30 +1,61 @@
...
+    """Fail if assets have the wrong tickers."""
+    example_base_currency = Ticker("USDT")
+    example_assets = {Ticker("USDT"): Money(Ticker("BTC"), 100)}
+
+    with pytest.raises(ValueError):
+        Portfolio(example_base_currency, example_assets)
+
+
+def test_reserved_assets(example_assets: dict) -> None:
     """Return free cash property."""
```
Comment (teadetime) [inline tests/backtest/test_portfolio.py:37]: naming here is a bit confusing for me

File: tests/backtest/test_portfolio.py [hunk 5]
```diff
@@ -1,30 +1,61 @@
...
-    example_cash = 10000.0
+    example_base_currency = Ticker("USDT")
+
+    p = Portfolio(example_base_currency, example_assets)
+
+    assert Money(Ticker("USDT"), 0) == p.reserved[Ticker("USDT")]
+
+
+def test_free_cash(example_assets: dict) -> None:
+    """Initialize correctly."""
```
Comment (teadetime) [inline tests/backtest/test_portfolio.py:46]: This docstring seems wrong?

Comment (teadetime) [reply tests/backtest/test_portfolio.py:46]: also `free_cash` seems like a weird name, is there something like available that suits better

File: tests/backtest/test_trade.py [hunk 6]
```diff
@@ -3,25 +3,25 @@
...
+    example_asset_pair: AssetPair,
     example_timestamp: dt.datetime,
 ) -> None:
     """Initialize correctly."""
     example_price = 100.0
     example_side = Side.BUY
 
     x = Trade(
-        symbol=example_symbol,
+        asset_pair=example_asset_pair,
```
Comment (teadetime) [inline tests/backtest/test_trade.py:18]: why no spaces between the equals?

File: tests/util/test_money.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+from coin_test.util import Money, Ticker
+
+
+def test_initialization() -> None:
+    """Initialize correctly."""
+    example_ticker = Ticker("BTC")
+    example_qty = 12.5
+
+    Money(example_ticker, example_qty)
```
Comment (teadetime) [inline tests/util/test_money.py:12]: should there be an assert here for the ticker and the qty

**Review by teadetime — APPROVED** [review]
These are mostly nits! looks good -- just some name changes

File: tests/util/test_money.py (hunk 7 above)
Comment (intermezzio) [reply tests/util/test_money.py:12]: Adding that now

File: tests/backtest/test_portfolio.py (hunk 4 above)
Comment (intermezzio) [reply tests/backtest/test_portfolio.py:37]: Renamed now

**Pull request merged**